cmd/client/main.go

```diff
@@ -209,7 +223,25 @@ func (c *Client) run(o *GrpcProxyClientOptions) error {
 	client := &http.Client{
 		Transport: transport,
 	}
-	requestURL := fmt.Sprintf("%s://%s:%d%s", o.requestProto, o.requestHost, o.requestPort, o.requestPath)
+
+	for i := 1; i <= o.testRequests; i++ {
+		err = c.makeRequest(o, client)
```

Comment:
We don't support reusing the same dialer for multiple requests. You'll need to obtain the dialer again for each request.

Reply:
I think this is a slightly tricky discussion. We have several separate layers here and we may be confusing them in this discussion. We have the outer tunnel (actually 2 separate tunnels, 1 from Client to Server, another from Server to Agent). We also a TCP connection on top of that tunnel (From  test client to test server) which is essentially a HTTP Session. On top of that we have individual HTTP requests. This is specifically HTTP 1.1 with session reuse. As long as we don't break the TCP connection (which is to say the HTTP session), then additional requests are just obfuscated traffic on top of that connection and the APIServer Network Proxy shouldn't care. That only works as long as we don't close the connection. There is a Close() in this method but its closing the response body (https://golang.org/src/net/http/transfer.go?s=97066:97143#L975) and not the connection, so everything should be fine. Its important that this work as things like exec and logs work on long lived HTTP sessions, so we need to handle it properly. If you try this test you will see that you can make multiple requests and everything works as longs as you don't kill the agent. In that case as you noted we should probably fail gracefully. Right now we just seem to hang.

Reply:
Thanks for the explanation. I should rephrase my original comment as "we do not support reusing the same dialer for multiple **connections**". With HTTP 1.1 Connection reuse, network proxy should indeed support subsequent requests since the original connection has not been closed. One thing I'd like to note is that golang's default `IdleConnTimeout` is 90 seconds (https://golang.org/src/net/http/transport.go?s=13608:13676) and k/k abides by this default. Thus, when running the test-client with a delay parameter greater than 90s (even when no agent restart is done) `--test-requests=2 --test-delay=100` the second request will fail 100% of the time because currently network proxy explicitly closes the tunnel when a close request is sent to to the inner TCP connection. I'm not sure if this is the desired behavior since I imagine we may have some cases of this kind of reuse in k/k. At the same time, there is no way for us to know whether a tunnel will be reused so to prevent leaking too many tunnel connections, we explicitly close the tunnel on the first CLOSE_RSP we obtain from the inner TCP connection (https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/master/konnectivity-client/pkg/client/client.go#L137).

As for the issue with the proxy-client hanging on agent restarts, that's a separate issue and I can submit a fix for that.

Reply:
Good point.

Added comment to this effect to the test-client code.